pubnub/models/subscription.py

```diff
@@ -0,0 +1,296 @@
+
+        tt = self.pubnub._subscription_manager._timetoken
+        if channel_list:
+            subscribe_operation = SubscribeOperation(
+                channels=self.get_subscribed_channels(),
+                channel_groups=self.get_subscribed_channel_groups(),
+                timetoken=tt,
+                presence_enabled=False
+            )
+            self.pubnub._subscription_manager.adapt_subscribe_builder(subscribe_operation)
+
+        return channel_list
+
+    def remove(self, subscription: Subscription) -> list:
+        channel_list = []
+        group_list = []
+
+        if subscription.subscription_type == PNSubscriptionType.CHANNEL:
+            for channel in subscription.subscription_names:
+                if channel in self.channels:
+                    self.channels[channel].remove(subscription)
+                    if self.channels[channel] == []:
+                        del self.channels[channel]
+                        channel_list.append(channel)
+        else:
+            for channel_group in subscription.subscription_names:
+                if channel_group in self.channel_groups:
+                    self.channel_groups[channel_group].remove(subscription)
+                if not self.channel_groups[channel_group]:
+                    del self.channel_groups[channel_group]
+                    group_list.append(channel_group)
+
+        if channel_list:
+            unsubscribe_operation = UnsubscribeOperation(
+                channels=channel_list,
+                channel_groups=group_list
+            )
+            self.pubnub._subscription_manager.adapt_unsubscribe_builder(unsubscribe_operation)
+        return channel_list
+
+    def get_subscribed_channels(self):
+        return list(self.channels.keys())
+
+    def get_subscribed_channel_groups(self):
+        return list(self.channel_groups.keys())
+
+    def get_subscriptions_for(self, subscription_type: PNSubscriptionType, name: str):
+        if subscription_type == PNSubscriptionType.CHANNEL:
+            return [channel for channel in self.get_subscribed_channels() if channel == name]
+        else:
+            return [group for group in self.get_subscribed_channel_groups() if group == name]
+
+    def get_all_listeners(self):
+        listeners = []
+
+        for channel in self.channels:
+            listeners += self.channels[channel]
+        for channel_group in self.channel_groups:
+            listeners += self.channel_groups[channel_group]
+        if self.global_listeners:
+            listeners += self.global_listeners
+        return set(listeners)
+
+    def add_listener(self, listener):
+        assert isinstance(listener, SubscribeCallback)
+        self.global_listeners.append(listener)
+
+    def remove_listener(self, listener):
+        assert isinstance(listener, SubscribeCallback)
+        self.global_listeners.remove(listener)
+
+    def unsubscribe_all(self):
```

Comment:
subscriptionset shouldn't have a `unsubscribe_all()` method, just `unsubscribe()` which unsubscribes all subscriptions that are part of that set

Reply:
You've seen the old version, before refactoring. In the current PR this is fixed. This method is now in `SubscriptionRegistry` an object to manage what is where